### ISSUE TITLE
Idea/Discussion: Do not associate a Request to an Event when creating Requests from API/CSV sources

### ISSUE THREAD

**Issue by mgmccarthy** [HTBox/allReady#1591]
Breaking the comments about this from #1590 and #1587 into this separate issue.

So far, the ideas that have been floated:
from @MisterJames in [1590](https://github.com/HTBox/allReady/issues/1590#issuecomment-264197849):
>  I'm not sure that I'm opposed to orphaned requests, but there are two ways that come to mind immediately to solve this.
 > 1. If we decide that orphaned requests are okay, we simply display a warning/alert at the top of every event page (for admins) and prompt them to "assign" the requests to the event. This could be done via a list with a filter set at the top/selection. This would be more work, but might allow finer-grained control.
> 2. If we prefer direct assignment, we could add an event drop down as part of the CSV import process. This would be trivial but would mean that all requests would have the same event assignment (which is likely okay).

with @stevejgordon giving a +1 to option 2

from @tonysurma in [1587](https://github.com/HTBox/allReady/issues/1587#issuecomment-264191816)
> @mgmccarthy interesting idea and then have the admin also have an 'associate' request just like the create request on the event. We could at least dump the requests under the organization at ingestion then so that Tony Org can't see requests that came in to Steve Org for example.

> Should we continue discussion here and change name (again :smile_cat: ) or move to another issue?

from @stevejgordon in [1587](https://github.com/HTBox/allReady/issues/1587#issuecomment-264198419)

> tonysurma Sounds sensible to have somewhere we can view "orphaned" requests. We could have a few issues and possibly address this at the code-a-thon. Something like

> Add Org level page (or maybe campaign level if we can at least determine campaign via API) to list requests without an event id set
Clicking on a request displays a page/modal allowing an campaign and event to be set
If that sounds reasonable I can break out some issues along those lines

> For CSV as it's a manual process to import I presume on the import screen we can/or should be able to choose the event we're importing under? With the assumption that data is validated by the admin prior to importing?

MY two cents:

- for @MisterJames option 2. This could be a good solution for CSV import, but will not work for API import
- for @tonysurma in his comment:  "We could at least dump the requests under the organization at ingestion then so that Tony Org can't see requests that came in to Steve Org for example".  The problem here is when we import a Request from CSV/API, we lack the context of Organizations/Campaigns/Events, etc... all we have to go on is the raw data in the Request. The most useful information to us is the geographic information we receive.
- same feedback for Steve's comment about being able to somehow associate an incoming Request with a Campaign.

What I think we should do here is give up on the association altogether. @stevejgordon said it best in [this comment](https://github.com/HTBox/allReady/issues/1587#issuecomment-264102398)
> The request in our domain represents an install location.

This boils down the Request to what it really is, a request for some type of service at a location (currently, we have city/state/zip to go on).

What I think we should do is:
- show ALL Requests in the "Unassigned Requests" area in each Event that do not have an Event/EventId assigned: since it sounds like admin-type users will be the ones assigning Requests to an Event, they will have the local knowledge about zip codes/zip code densities/general knowledge on the area to make an assignment based on city/state/zip in their local area that makes sense.
![image](https://cloud.githubusercontent.com/assets/2428993/20802244/77fa8266-b7b9-11e6-9c38-af6cb2058245.png)

![image](https://cloud.githubusercontent.com/assets/2428993/20802255/84e82596-b7b9-11e6-8d95-b0f245093f00.png)

- for Requests that do have an Event associated with them, keep showing those Requests in the Unassigned Requests area of the associated Event (this is how the code works not). This will allow the current create Request UI to continue functioning as normal. If a user wants to create a Request in the context of an Event (aka, via the UI), and they type in a zip code that does not make sense for the Event, then that's on them. Sounds like we're not going to stop them.

Basically, put the onus on the user of the UI that is assigning Requests to an Event to make an assignment that makes sense based on city/state/zip.

**Comment by tonysurma** [comment]
I am good with the proposal except - once we get to more than one org submitting API requests we can't have Tony Org + Steve Org seeing the same pool of requests.  Couldn't we use the auth token to see which user has that auth token and then what org they are in?

However that is a v1.1 issue and not a v1 issue so for now I am happy to have them all come in and just be tagged with the organization for the redcross and that will be fine and then the rest can generally flow as documented (I skimmed above so need to read all details but wanted to get a quick direction out there)

**Comment by mgmccarthy** [comment]
@tonysurma, I think I understand now.

When you're talking about Orgs, you're saying we should represent each external source (in this case, GASA/Red Cross) with an internal Organization in allReady?

If that's the case, what we'll really need is a unique id we'll assign to each org we'll be receiving Requests from, and THAT's why you're saying we could use a token for that identification. 

That token would be sent with every Request, and on our end, we could correlate the token to the owning Org, and that would be the start of how we could "corral" these Requests into some type of owndership with allReady.

Regardless of the how we technically create that association (use tokens/claims or write our own way to manage/handle these unique id's), the goal is the same. All incoming Requests should at least be able to be routed to the appropriate Org on allReady's side.

Do I understand it right?

**Comment by tonysurma** [comment]
quick answer: right and we already have organizations in our system so just need to get it mapped even if you hard code it for December 😈

**Comment by mgmccarthy** [comment]
So, I think adding the Event drop-down to the Import Request CSV screen to associate those Requests with an event is a good idea.

For the Request API imports, for the December release, I MIGHT be doing what @tonysurma mentions above:
> right and we already have organizations in our system so just need to get it mapped even if you hard code it for December 

since we're only launching for GASA/Red Cross, this approach is entirely feasible for the December release. If we go this route, we'll still need a way to look at these Requests by Org, which @stevejgordon  has already proposed a UI that would allow us to do so, then take those Requetss and associate them to an Event.

**Comment by mgmccarthy** [comment]
@tonysurma  @stevejgordon  One note about routing incoming Requests to an Org based off of a token... 

this solution requires the Org to exist in allReady before we can start to accept and process API requests. I'm not too sure if this is a "big deal" or not, but it does require some setup on the web/UI end before we can officially start to test the import.

Did you want me to seed a new Org for Red Cross (or GASA?)... not too sure what to call it...

or did you want it to be a manual process where the user has to create the org?

**Comment by tonysurma** [comment]
quick note: we already have a set of UI for creating orgs and associating users and I *think* the user token generated is tied to a user so it should be tied to an org.

For dev you could use the seeded htbox org.

For the hard coding, perhaps we can do a config value of which ID that I can set in production release from whatever org they get in the database.

**Comment by mgmccarthy** [comment]
@tonysurma, @MisterJames  is currently working on this:

> we already have a set of UI for creating orgs and associating users and I think the user token generated is tied to a user so it should be tied to an org.

right now, the token that is being generated for a user is not being recognized on the inbound trip back to our endpoint. But you are correct, the token is assigned to a user, and then that user will be associated to an Org.

One other problem I just thought of:

How do we communicate the status of this request back to GASA's endpoint if we're not going to be able to make a decision at the point we receive the request whether or not we can service the Request?

If we can agree that a "serviceable request" is if the incoming Request can be mapped to an existing Org, we can still invoke GASA API endpoint with true/false, so it's a non-issue. But if we need to wait for the Request to be assigned to an Itinerary on an Event, then the communication back to GASA's API would be deferred, or in the case the Request is never assigned, GASA would never hear back from us.

**Comment by mgmccarthy** [comment]
@tonysurma, I'm going to add an OrganizationId to the Request model and for now, we'll hard code all incoming API requests to that OrgId. 

Eventually, we should use our generated token we give to the provider that will be in each incoming request to look up the user's OrgId based on the token, and assign the OrgId to that Request (v1.1). 

For now, we can build a UI that will allow a user to look at all the Requests in that given organization that are not associated with an Event. This UI should allow the user to assign these Requests to Events.

As far as reporting whether or not we will service the Request back to GASA's API, I'm going to make the assumption that if we can match an incoming request's token with a user/user's organization, that means we can service the request.

With the "hard-coded" approach for v1, that would mean we can service every request they send us.

If this if not correct, let me know. 

For instance, just b/c we assign an OrgId to a Request might not mean we can service it. We might only be able to communicate back to GASA's API that we can service a request once the Request has been associated to an Event.

If this is the case, then we would report false back to GASA for all requests imported from them, and eventually, they would receive true's when we associate that Request to an Event.

Thanks.

**Comment by stevejgordon** [comment]
I've added three stages of work for the code-a-thon (hopefully) that will enable the initial pieces for the v1 functionality above.

**Comment by mgmccarthy** [comment]
@tonysurma @MisterJames @stevejgordon it will be important to figure out soon: 

1. how can we update GASA's endpoint with the correct status (we service or do not service the Request) now that we'll be associating all API Requests with the red cross org, and then building out a UI to associate requests to events. Part of the answer I'm waiting on GASA for, b/c I need to know how flexible they are with us reporting status back to them

2. getting the user token situation in Issue #1586 solved. We, at minimum, need to use that token value as an authorization mechanism for incoming API Requests.

Thanks!

**Comment by EmilyLuijbregts** [comment]
@MisterJames @tonysurma @stevejgordon  Just wanting to follow up on this task and the question from @mgmccarthy. Do we need to have a chat to discuss the two points?

**Comment by tonysurma** [comment]
thanks for th eping @EmilyLuijbregts 

(2) is in progress or at least planned and (1) we resolved in a discussion on their repo.  We will report true if it matches a list of region id's that we are supporting 'hardcoded' to start and false otherwise and only 'save' those that match our region ids.

**Comment by mgmccarthy** [comment]
@EmilyLuijbregts, the functionality outlined above in this issue for API Requests have actually been decided on for v1 release pending a working Issue #1586. Even if we don't get Issue #1586 working, we have a fall back plan for the launch. The UI is currently being worked on and that work is reflected in Issues #1596 and #1597

As far as associating OrgId to CSV import Requests, I'm too sure where we are with that one? I know the work is outlined in Issue #1590 and some above, but I don't think we have an explicit task to represent that work yet.

@tonysurma @MisterJames @stevejgordon please chime in if what I just said is not true.
